prow/kube/prowjob_collector.go

```diff
@@ -0,0 +1,253 @@
+	//prometheus.DescribeByCollect(pjc, ch)
+	// Normally, we'd send descriptors into the channel. However, we cannot do so for these
+	// metrics as their label sets are dynamic. This is a take-our-own-risk action and also a
+	// compromise for implementing a metric with both dynamic keys and dynamic values in
+	// the label set.
+	// https://godoc.org/github.com/prometheus/client_golang/prometheus#hdr-Custom_Collectors_and_constant_Metrics
+}
+
+func (pjc prowJobCollector) Collect(ch chan<- prometheus.Metric) {
+	logrus.Debug("ProwJobCollector collecting ...")
+	prowJobs, err := pjc.lister.List(metav1.ListOptions{LabelSelector: pjc.selector})
+	if err != nil {
+		logrus.WithError(err).Errorf("Failed to list prow jobs with selector '%s'", pjc.selector)
+		return
+	}
+	for _, pj := range prowJobs {
+		if pj.Spec.Agent != pjc.agent {
+			continue
+		}
+		pjLabelKeys, pjLabelValues := kubeLabelsToPrometheusLabels(pj.Labels)
+		pjLabelKeys = append([]string{"prow_job_name", "prow_job_namespace"}, pjLabelKeys...)
+		pjLabelValues = append([]string{pj.Name, pj.Namespace}, pjLabelValues...)
+		desc := prometheus.NewDesc(
+			"prow_job_labels",
+			"Kubernetes labels converted to Prometheus labels.",
+			pjLabelKeys, nil,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			desc,
+			prometheus.GaugeValue,
+			// always 1 since there is only 1 prow job for each namespace and each prow job name
+			float64(1),
+			pjLabelValues...,
+		)
+	}
+}
+
+func registerProwJobCollector(name string, lister lister, selector string, agent prowapi.ProwJobAgent, reg prometheus.Registerer) *prowJobCollector {
+	c := &prowJobCollector{
+		name:     name,
+		lister:   lister,
+		selector: selector,
+		agent:    agent,
+	}
+	prometheus.WrapRegistererWith(prometheus.Labels{"collector_name": name}, reg).MustRegister(c)
+	return c
+}
+
+var (
+	invalidLabelCharRE    = regexp.MustCompile(`[^a-zA-Z0-9_]`)
+	escapeWithDoubleQuote = strings.NewReplacer("\\", `\\`, "\n", `\n`, "\"", `\"`)
+)
+
+// https://github.com/kubernetes/kube-state-metrics/blob/1d69c1e637564aec4591b5b03522fa8b5fca6597/internal/store/utils.go#L60
+func kubeLabelsToPrometheusLabels(labels map[string]string) ([]string, []string) {
+	labelKeys := make([]string, 0, len(labels))
+	for k := range labels {
+		labelKeys = append(labelKeys, k)
+	}
+	sort.Strings(labelKeys)
+
+	labelValues := make([]string, 0, len(labels))
+	for i, k := range labelKeys {
+		labelKeys[i] = "label_" + sanitizeLabelName(k)
+		labelValues = append(labelValues, escapeString(labels[k]))
+	}
+	return labelKeys, labelValues
+}
+
+func sanitizeLabelName(s string) string {
+	return invalidLabelCharRE.ReplaceAllString(s, "_")
+}
+
+// https://github.com/kubernetes/kube-state-metrics/blob/1d69c1e637564aec4591b5b03522fa8b5fca6597/pkg/metric/metric.go#L96
+func escapeString(v string) string {
+	return escapeWithDoubleQuote.Replace(v)
+}
+
+const (
+	pjControllerName = "prowjob-controller-for-collector"
+)
+
+type PJController struct {
```

Comment:
Could you explain what this is supposed to do? From what I can see it just appends Prowjobs to a list on update and create events. This does not currently work because you pass in a list instead of a pointer to it but if it would, it would be a huge memory leak.

Also, I recommend using controller-runtime for controllers, it allows you to avoid a lot of the boilerplate. A minimal example is e.G. this: https://github.com/kubernetes-sigs/cluster-api/blob/c124348ca703fcd826a36d0b85faa0ae19ee861c/controllers/machine_controller.go#L58-L78 (and if you have overall questions to that, I am happy to make a small introduction)

Reply:
Will take a look and learn. Thanks a lot. @alvaroaleman

Reply:
Yes. Passing a list and appending it are indeed a problem. The appending out of the function [has no impact](https://nanxiao.gitbooks.io/golang-101-hacks/content/posts/pass-slice-as-a-function-argument.html). Let me fix it during the runtime refactoring.

Reply:
From the first glance, the example is not complete. :)

`func (r *MachineReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options)`: figuring out how to get those args.

Reply:
Changed to pointer of the list.

@alvaroaleman My questions about the `controller-runtime`: Would it help the memory issue? What is the diff from the current lister/informer API? Links of its doc would be helpful.

Reply:
controller-runtime won't change anything about the memory usage, it still uses the same machinery underneath as an ordinary controller. The main advantage is that you don't need a lot of the boilerplate, resulting in less code to write, review and maintain.

Circling back to the original question thought: What problem is this controller supposed to solve? From what I can see it only appends to a list, but never does anything with that list or cleans it up. Consequently, it will leak vast amounts of memory but not really do anything useful.

Reply:
Ideally, it is desired to have a controller which can list the existing prow jobs in the system.
We do not modify any prow job. The existing controller in plank would do the job if we do not consider efficiency.

https://github.com/kubernetes/test-infra/blob/4cd31c1afe10ff3ed0cca2658cf9257825101d18/prow/cmd/plank/main.go#L126

We introduce informer-lister controller because [it is more efficient than listing all prow-jobs on each scraping by prometheus servers](https://github.com/kubernetes/test-infra/pull/14457#discussion_r328874505). Let us put aside for the moment if informer pattern is sufficient for this scenario (see https://github.com/kubernetes/test-infra/pull/14457#discussion_r329332698).

When prometheus does scraping, it will call `Collect(ch chan<- prometheus.Metric)`  which depends on the `lister` of the controller.

https://github.com/kubernetes/test-infra/blob/bb2fbfe164bcfc7894cb6b6051ee7fe18a33c20f/prow/kube/prowjob_collector.go#L62-L65

The useful part of the controller for scraping is the lister and we need nothing else for generating the prometheus metric samples.

Reply:
I see. So the only part you actually need is the lister and there is no need for a controller. Listers are bulding blocks of controllers but not the other way around, i.E. you can use the former without needing the latter.

Reply:
Removed the added controller.
Only the lister is used here and the lister is from the informer factory because we still want to utilize the cache which the existing controller did not have.